tests/unit/test_tls_lib.py

```diff
@@ -0,0 +1,311 @@
+
+        #  Tests for non-leader unit (ie external certificates)
+        self.harness.set_leader(False)
+        action_event = mock.Mock()
+        action_event.params = {}
+        app_rsa_key = self.harness.charm.app_peer_data["key"]
+        app_csr = self.harness.charm.app_peer_data["csr"]
+
+        # generated rsa key test - non-leader
+        self.harness.charm.tls._on_set_tls_private_key(action_event)
+        self.verify_external_rsa_csr()
+        # non-leaders should not reset the app key and app csr
+        self.verify_internal_rsa_csr(
+            specific_rsa=True, expected_rsa=app_rsa_key, specific_csr=True, expected_csr=app_csr
+        )
+
+        # provided rsa key test - non-leader
+        set_unit_rsa_key = self.harness.charm.unit_peer_data["key"]
+        action_event.params = {"external-key": set_unit_rsa_key}
+        self.harness.charm.tls._on_set_tls_private_key(action_event)
+        self.verify_external_rsa_csr(specific_rsa=True, expected_rsa=set_unit_rsa_key)
+        # non-leaders should not reset the app key and app csr
+        self.verify_internal_rsa_csr(
+            specific_rsa=True, expected_rsa=app_rsa_key, specific_csr=True, expected_csr=app_csr
+        )
+
+    @patch_network_get(private_address="1.1.1.1")
+    def test_tls_relation_joined_non_leader(self):
+        """Test that non-leader units set only external certificates."""
+        self.harness.set_leader(False)
+        self.relate_to_tls_certificates_operator()
+        # non leaders should not be allowed to set internal certificates
+        self.verify_internal_rsa_csr(
+            specific_rsa=True, expected_rsa=None, specific_csr=True, expected_csr=None
+        )
+        self.verify_external_rsa_csr()
+
+    @patch_network_get(private_address="1.1.1.1")
+    def test_tls_relation_joined_leader(self):
+        """Test that leader units set both external and internal certificates."""
+        self.harness.set_leader(True)
+        self.relate_to_tls_certificates_operator()
+        self.verify_internal_rsa_csr()
+        self.verify_external_rsa_csr()
+
+    @patch("charms.mongodb_libs.v0.mongodb_tls.restart_mongod_service")
+    @patch_network_get(private_address="1.1.1.1")
+    def test_tls_relation_broken_non_leader(self, restart_mongod_service):
+        """Test non-leader removes only external cert & chain."""
+        # set initial certificate values
+        self.harness.set_leader(True)
+        rel_id = self.relate_to_tls_certificates_operator()
+        app_rsa_key = self.harness.charm.app_peer_data["key"]
+        app_csr = self.harness.charm.app_peer_data["csr"]
+
+        self.harness.set_leader(False)
+        self.harness.remove_relation(rel_id)
+        self.assertEqual(self.harness.charm.unit_peer_data.get("ca", None), None)
+        self.assertEqual(self.harness.charm.unit_peer_data.get("cert", None), None)
+        self.assertEqual(self.harness.charm.unit_peer_data.get("chain", None), None)
```

Comment:
You could use `self.assertIsNone` here, and you can also use self.charm instead of self.harness.charm.

Reply:
TWIL thanks @WRFitch